packages/fetchai/connections/http/connection.py

```diff
@@ -0,0 +1,565 @@
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""HTTP connection, channel, server, and handler"""
+
+import asyncio
+import logging
+from asyncio import CancelledError
+from http.server import BaseHTTPRequestHandler, HTTPServer
+from threading import Lock, Thread
+from typing import Dict, Optional, Set, cast
+from uuid import uuid4
+
+from openapi_core import create_spec
+from openapi_core.validation.request.datatypes import (
+    OpenAPIRequest,
+    RequestParameters,
+)
+from openapi_core.validation.request.shortcuts import validate_request
+from openapi_core.validation.request.validators import RequestValidator
+
+from openapi_spec_validator.schemas import read_yaml_file
+
+from werkzeug.datastructures import ImmutableMultiDict
+
+from aea.configurations.base import ConnectionConfig, PublicId
+from aea.connections.base import Connection
+from aea.mail.base import Address, Envelope, EnvelopeContext, URI
+
+from packages.fetchai.protocols.http.message import HttpMessage
+from packages.fetchai.protocols.http.serialization import HttpSerializer
+
+SUCCESS = 200
+NOT_FOUND = 404
+REQUEST_TIMEOUT = 408
+
+logger = logging.getLogger(__name__)
+
+RequestId = str
+
+
+class Request(OpenAPIRequest):
+    """Generic request object."""
+
+    @property
+    def id(self) -> RequestId:
+        return self._id
+
+    @id.setter
+    def id(self, id: RequestId) -> None:
+        self._id = id
+
+    @classmethod
+    def create(cls, request_handler: BaseHTTPRequestHandler) -> "Request":
+        method = request_handler.command.lower()
+
+        # gets deduced by path finder against spec
+        path = {}  # type: Dict
+
+        url = "http://{}:{}{}".format(
+            *request_handler.server.server_address, request_handler.path
+        )
+
+        content_length = request_handler.headers["Content-Length"]
+        body = (
+            None
+            if content_length is None
+            else request_handler.rfile.read(int(content_length)).decode()
+        )
+
+        mimetype = request_handler.headers.get_content_type()
+        parameters = RequestParameters(
+            query=ImmutableMultiDict(request_handler.headers.get_params()),
+            header=request_handler.headers,
+            path=path,
+        )
+        request = Request(
+            full_url_pattern=url,
+            method=method,
+            parameters=parameters,
+            body=body,
+            mimetype=mimetype,
+        )
+        request.id = uuid4().hex
```

Comment:
Doing this via a property was preferred after trying several attempts of satisfying mypy.